### ISSUE TITLE
singularity.spec missing "BuildRequires: python"

### ISSUE THREAD

**Issue by tabaer** [apptainer/singularity#853]
## Version of Singularity:

2.3.1

### Expected behavior

"mock <args> --rebuild $SRPM" builds RPMs.

### Actual behavior

"mock <args> --rebuild $SRPM" fails:

---
mock -r rhel-7-x86_64 --resultdir=${DIR}/built --configdir=${DIR}/../mock/${HOST} --rebuild $SRPM --cleanup-after
[...]
+ ./configure --build=x86_64-redhat-linux-gnu --host=x86_64-redhat-linux-gnu --program-prefix= --disable-dependency-tracking --prefix=/usr --exec-prefix=/usr --bindir=/usr/bin --sbindir=/usr/sbin --sysconfdir=/etc --datadir=/usr/share --includedir=/usr/include --libdir=/usr/lib64 --libexecdir=/usr/libexec --localstatedir=/var --sharedstatedir=/var/lib --mandir=/usr/share/man --infodir=/usr/share/info --without-slurm
checking build system type... x86_64-redhat-linux-gnu
checking host system type... x86_64-redhat-linux-gnu
checking target system type... x86_64-redhat-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for style of include used by make... GNU
checking for x86_64-redhat-linux-gnu-gcc... no
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... none
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking minix/config.h usability... no
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking how to convert x86_64-redhat-linux-gnu file names to x86_64-redhat-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-redhat-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for x86_64-redhat-linux-gnu-objdump... no
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for x86_64-redhat-linux-gnu-dlltool... no
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for x86_64-redhat-linux-gnu-ar... no
checking for ar... ar
checking for archiver @FILE support... @
checking for x86_64-redhat-linux-gnu-strip... no
checking for strip... strip
checking for x86_64-redhat-linux-gnu-ranlib... no
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for x86_64-redhat-linux-gnu-mt... no
checking for mt... no
checking if : is a manifest tool... no
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for x86_64-redhat-linux-gnu-gcc... gcc
checking whether we are using the GNU C compiler... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to accept ISO C89... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking dependency style of gcc... (cached) none
checking dependency style of gcc... (cached) none
checking for namespace: CLONE_NEWPID... yes
checking for namespace: CLONE_PID... no
checking for namespace: CLONE_FS... yes
checking for namespace: CLONE_NEWNS... yes
checking for namespace: CLONE_NEWUSER... yes
checking for namespace: CLONE_NEWIPC... yes
checking for feature: NO_NEW_PRIVS... yes
checking for feature: MS_SLAVE... yes
checking for feature: MS_REC... yes
checking for feature: MS_PRIVATE... yes
checking for setns... yes
checking for overlayfs... maybe
checking for python... no
configure: error: Please install python before installing.

RPM build errors:
error: Bad exit status from /var/tmp/rpm-tmp.KqpTsq (%build)
    Bad exit status from /var/tmp/rpm-tmp.KqpTsq (%build)
ERROR: Exception(SRPMS/singularity-2.3.1-0.1.el7.src.rpm) Config(rhel-7-x86_64) 0 minutes 17 seconds
INFO: Results and/or logs in: /users/sysp/troy/src/private/osc-rpmbuilds/singularity/built
INFO: Cleaning up build root ('cleanup_on_failure=True')
Start: clean chroot
Finish: clean chroot
ERROR: Command failed. See logs for output.
 # bash --login -c /usr/bin/rpmbuild -bb --target x86_64 --nodeps /builddir/build/SPECS/singularity.spec
---

### Steps to reproduce behavior

```
#!/bin/bash
DIR=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
HOST=${HOSTNAME%%.*}
VERSION=2.3.1
PREFIX=/opt/singularity

# download tarball, if necessary
if [ ! -d SOURCES ]; then
  mkdir SOURCES
fi
if [ ! -e SOURCES/singularity-$VERSION.tar.gz ]; then
  cd SOURCES
  wget "https://github.com/singularityware/singularity/releases/download/$VERSION/singularity-$VERSION.tar.gz"
  cd ..
fi

tar xzvf SOURCES/singularity-$VERSION.tar.gz singularity-$VERSION/singularity.spec
cp singularity-$VERSION/singularity.spec .

# BUILD RHEL7
SRPM_OUT=$(rpmbuild -bs --define "_topdir ${DIR}" --define "rhel 7" --define "dist .el7" singularity.spec)
SRPM=SRPMS/$(echo $SRPM_OUT | awk '{ print $2 }' | xargs basename)
mock -r rhel-7-x86_64 --resultdir=${DIR}/built --configdir=${DIR}/../mock/${HOST} --rebuild $SRPM --cleanup

```

Adding the following to singularity.spec causes the build to succeed:

---

BuildRequires: python

---

**Comment by tmozgach** [comment]
I have the same error even I have the python! The system is Ubuntu!

**Comment by vsoch** [comment]
I think this was an issue in the `singularity.in` spec file - does it occur when you use development branch?

**Comment by tmozgach** [comment]
@vsoch The `python` has been recently updated and a symlink solves  the problem:

``` ln -s python2 /home/linuxbrew/.linuxbrew/bin/python ```


But I got the following the error: 

```
linuxbrew@7629ddeac9f9:~/singularity-2.3.1$ singularity create centos7.img
Initializing Singularity image subsystem
Opening image file: centos7.img
Creating 768MiB image
Binding image to loop
Singularity: create-suid (U=0,P=18542)> No more available loop devices, try increasing 'max loop devices' in singularity.conf

ERROR  : No more available loop devices, try increasing 'max loop devices' in singularity.conf
Singularity: create-suid (U=0,P=18542)> Retval = 255

ABORT  : Retval = 255

```


My system is `Ubuntu 14.04.5 LTS`

**Comment by verdurin** [comment]
This is still broken in `2.4`; appears fixed in `master`.

I've setup a COPR in Fedora for downloadable RPMs:

https://copr.fedorainfracloud.org/coprs/verdurin/singularity/

which includes the `BuildRequires: python` fix.

**Comment by vsoch** [comment]
@verdurin @tmozgach if it's fixed in master it should go out with 2.4.1 then.

**Comment by dtrudg** [comment]
`BuildRequires: python` is in the spec file now in later 2.4 releases.
